File: guilib-bukkit/src/main/java/net/developertobi/guilib/bukkit/api/gui/GuiHandlerImpl.java
package net.developertobi.guilib.bukkit.api.gui;

import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.Multimap;
import net.developertobi.guilib.api.gui.Gui;
import net.developertobi.guilib.api.gui.GuiController;
import net.developertobi.guilib.api.gui.GuiHandler;
import net.developertobi.guilib.api.gui.GuiProperties;
import net.developertobi.guilib.api.gui.GuiView;
import net.developertobi.guilib.api.utils.MathUtils;
import net.developertobi.guilib.bukkit.GuiLibBukkit;
import net.kyori.adventure.text.Component;
import org.bukkit.Bukkit;
import org.bukkit.NamespacedKey;
import org.bukkit.entity.Player;
import org.bukkit.persistence.PersistentDataType;

public class GuiHandlerImpl implements GuiHandler {

    private static NamespacedKey openGuiKey;

    private static NamespacedKey getOpenGuiKey() {
        if (openGuiKey == null) {
            openGuiKey = new NamespacedKey(GuiLibBukkit.getInstance(), "open-gui");
        }
        return openGuiKey;
    }

    private final Multimap<Player, GuiView> inventories = ArrayListMultimap.create();

    @Override
    public Multimap<Player, GuiView> getInventories() {
        return inventories;
    }

    @Override
    public void openStaticView(Player holder, Component title, Gui provider, boolean forceSyncOpening) {
        GuiView gui = cacheView(holder, title, provider, true);
        if (gui == null) return;
        gui.open(holder, forceSyncOpening);
    }

    @Override
    public GuiView cacheView(Player holder, Component title, Gui provider, boolean staticGui) {
        GuiProperties annotation = provider.getClass().getAnnotation(GuiProperties.class);
        String permission = annotation == null ? "" : annotation.permission();

        if (permission != null && !permission.isEmpty() && !holder.hasPermission(permission)) {
            return null;
        }

        GuiController controller = new GuiControllerImpl(provider);
        org.bukkit.inventory.Inventory rawInventory = Bukkit.createInventory(holder, controller.getRows() * controller.getColumns(), title);

        controller.setRawInventory(rawInventory);

        for (int i = 0; i < controller.getSlotCount(); i++) {
            controller.getContents().put(MathUtils.slotToPosition(i, controller.getColumns()), null);
        }

        provider.init(holder, controller);

        controller.updateRawInventory();

        GuiView gui = new GuiImpl(provider, title, controller, staticGui);

        inventories.put(holder, gui);
        return gui;
    }

    @Override
    public void updateCachedView(Player holder, String viewId) {
        GuiView gui = getView(holder, viewId);
        if (gui == null) return;

        Component title = gui.getTitle();
        Gui provider = gui.getController().getProvider();

        inventories.get(holder).removeIf(inv -> inv.getName().equalsIgnoreCase(viewId));
        cacheView(holder, title, provider);

        GuiView elytraView = getView(holder, viewId);
        if (elytraView == null) return;
        String key = elytraView.getName();

        String openGuiName = holder.getPersistentDataContainer().get(getOpenGuiKey(), PersistentDataType.STRING);
        if (openGuiName == null || !key.equalsIgnoreCase(openGuiName)) return;

        elytraView.open(holder, true);
    }

    @Override
    public void clearCachedViews(Player holder) {
        inventories.removeAll(holder);
        if (holder.getPersistentDataContainer().has(getOpenGuiKey(), PersistentDataType.STRING)) {
            holder.getPersistentDataContainer().remove(getOpenGuiKey());
        }
    }

    @Override
    public void removeCachedView(Player holder, GuiView view) {
        inventories.remove(holder, view);
        if (holder.getPersistentDataContainer().has(getOpenGuiKey(), PersistentDataType.STRING)) {
            holder.getPersistentDataContainer().remove(getOpenGuiKey());
        }
    }

    @Override
    public GuiView getView(Player holder, String name) {
        return inventories.get(holder).stream()
                .filter(inv -> inv.getName().equalsIgnoreCase(name))
                .findFirst()
                .orElse(null);
    }
}

File: guilib-bukkit/bin/main/net/developertobi/inventorylib/bukkit/api/inventory/GuiHandlerImpl.kt
package net.developertobi.inventorylib.bukkit.api.inventory

import com.google.common.collect.ArrayListMultimap
import com.google.common.collect.Multimap
import net.kyori.adventure.text.Component
import net.developertobi.inventorylib.api.inventory.GuiHandler
import net.developertobi.inventorylib.api.inventory.GuiProperties
import net.developertobi.inventorylib.api.inventory.Gui
import net.developertobi.inventorylib.api.inventory.GuiView
import net.developertobi.inventorylib.api.utils.MathUtils
import net.developertobi.inventorylib.bukkit.GuiInventoryBukkit
import org.bukkit.Bukkit
import org.bukkit.NamespacedKey
import org.bukkit.entity.Player
import org.bukkit.persistence.PersistentDataType

class GuiHandlerImpl : GuiHandler {

    override val inventories: Multimap<Player, GuiView> = ArrayListMultimap.create()

    companion object {
        private val OPEN_INVENTORY_KEY: NamespacedKey by lazy {
            NamespacedKey(GuiInventoryBukkit.instance, "open-inventory")
        }
    }

    override fun openStaticView(holder: Player, title: Component, provider: Gui, forceSyncOpening: Boolean) {
        val inventory: GuiView = cacheView(holder, title, provider, true) ?: return
        inventory.open(holder, forceSyncOpening)
    }

    override fun cacheView(holder: Player, title: Component, provider: Gui) {
        cacheView(holder, title, provider, false)
    }

    override fun cacheView(holder: Player, title: Component, provider: Gui, staticInventory: Boolean): GuiView? {
        val permission = if (provider.javaClass.getAnnotation(GuiProperties::class.java) == null) ""
        else provider.javaClass.getAnnotation(GuiProperties::class.java).permission

        if (!permission.equals("", true) && !holder.hasPermission(permission)) return null

        val controller: net.developertobi.inventorylib.api.inventory.GuiController = GuiControllerImpl(provider)
        val rawInventory = Bukkit.createInventory(holder, controller.getRows() * controller.getColumns(), title)

        controller.rawInventory = rawInventory

        for (i in 0 until controller.inventorySlotCount) {
            controller.contents[MathUtils.slotToPosition(i, controller.getColumns())] = null
        }

        provider.init(holder, controller)

        controller.updateRawInventory()

        val inventory: GuiView = GuiInventoryImpl(provider, title, controller, staticInventory)

        inventories.put(holder, inventory)
        return inventory
    }

    override fun updateCachedView(holder: Player, viewId: String) {
        val inventory: GuiView = getView(holder, viewId) ?: return

        val title: Component = inventory.title
        val provider: Gui = inventory.controller.provider

        inventories[holder].removeIf { inv: GuiView -> inv.name.equals(viewId, true) }
        cacheView(holder, title, provider)

        val elytraView = getView(holder, viewId) ?: return
        val key: String = elytraView.name

        val openInventoryName =
            holder.persistentDataContainer.get(OPEN_INVENTORY_KEY, PersistentDataType.STRING) ?: return
        if (!key.equals(openInventoryName, true)) return

        elytraView.open(holder, true)
    }

    override fun clearCachedViews(holder: Player) {
        inventories.removeAll(holder)
        if (holder.persistentDataContainer.has(OPEN_INVENTORY_KEY, PersistentDataType.STRING))
            holder.persistentDataContainer.remove(OPEN_INVENTORY_KEY)
    }

    override fun removeCachedView(holder: Player, view: GuiView) {
        inventories.remove(holder, view)
        if (holder.persistentDataContainer.has(OPEN_INVENTORY_KEY, PersistentDataType.STRING)) {
            holder.persistentDataContainer.remove(OPEN_INVENTORY_KEY)
        }
    }

    override fun getView(holder: Player, name: String): GuiView? {
        return inventories[holder].first { it.name.equals(name, true) }
    }
}

